File: 146. LRU Cache/solution.py
class Node:
    def __init__(self, key, val):
        self.key = key
        self.val = val
        self.next = None
        self.prev = None
        
class LRUCache:
    def __init__(self, capacity: int):
        self.cap = capacity
        self.head = Node(None, None)
        self.tail = Node(None, None)
        self.head.prev = self.tail
        self.tail.next = self.head
        self.map = {}

    def get(self, key: int) -> int:
        head = self.head
        tail = self.tail
        
        if key in self.map:
            node = self.map[key]
            node.prev.next = node.next
            node.next.prev = node.prev
            node.prev = head.prev
            head.prev = node
            node.prev.next = node
            node.next = head
            return node.val
        else:
            return -1
        
    def put(self, key: int, value: int) -> None:
        head = self.head
        tail = self.tail

        node = None    
        if key in self.map:
            node = self.map[key]
            node.val = value
            node.prev.next = node.next
            node.next.prev = node.prev
        
        if not node:
            if not self.cap:
                node = tail.next
                self.map.pop(node.key)
                node.prev.next = node.next
                node.next.prev = node.prev
            else:
                self.cap -= 1
            node = Node(key, value)
            self.map[key] = node
        
        node.prev = head.prev
        head.prev = node
        node.prev.next = node
        node.next = head

        
# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: 146. LRU Cache/solution.java
class LRUCache {
    private int cap;
    private HashMap<Integer, Record> map;
    private Record tail;
    private Record head;
    
    public LRUCache(int capacity) {
        cap = capacity;
        tail = new Record(-1, -1);
        head = new Record(-1, -1);
        head.prev = tail;
        tail.next = head;
        map = new HashMap<Integer, Record>();
    }
    
    public int get(int key) {
        if (map.containsKey(key)) {
            Record record = map.get(key);
            Record prev, next;
            prev = record.prev;
            next = record.next;
            prev.next = next;
            next.prev = prev;
            
            prev = head.prev;
            next = head;
            record.prev = prev;
            record.next = next;
            next.prev = record;
            prev.next = record;
            
            return record.value;
        } else {
            return -1;
        }
    }
    
    public void put(int key, int value) {
        Record record;
        Record prev, next;
        if (map.containsKey(key)) {
            record = map.get(key);
            prev = record.prev;
            next = record.next;
            prev.next = next;
            next.prev = prev;
            record.value = value;
        } else if (cap == 0) {
            record = tail.next;
            map.remove(record.key);
            prev = record.prev;
            next = record.next;
            prev.next = next;
            next.prev = prev;

            record.key = key;
            record.value = value;
            map.put(key, record);
        } else {
            cap--;
            record = new Record(key, value);
            map.put(key, record);
        }
        
        prev = head.prev;
        next = head;
        record.prev = prev;
        record.next = next;
        next.prev = record;
        prev.next = record;
    }
    
    class Record {
        public int key;
        public int value;
        public Record next;
        public Record prev;
        
        public Record(int k, int v) {
            key = k;
            value = v;
            next = null;
            prev = null;
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */
